doc/developer/design/20230315_backup_restore.md

```diff
@@ -0,0 +1,210 @@
+
+The proposed CRDB backup strategy may lead to an unrestorable backup if we experience a causal reverse. However, we expect this to be rare and straightforward to work around when it does occur.
+
+## Alternatives to single-shard backups
+
+### Copying-based approaches
+
+One natural way to make a backup of a shard would be to copy out a state rollup and all the files it recursively references out to some external location. This is straightforward to implement using the existing blob and consensus APIs and requires no new infrastructure.
+
+- PUTs in S3 are (relatively) very expensive, and GETs marginally less so. (On the other hand, we spend comparatively very little on storage.) Copying out a significant percentage of the files we write would push up our S3 costs in a way that leaving data in place does not.
+- Somebody has to trigger the copying! If `environmentd` is coordinating the environment-wide backup process, making it copy every file in every shard is lots of extra work. (Though thankfully S3 can copy files without transferring all the data through the client.) We might need to introduce an additional sidecar just for backups, which is significant operational burden.
+
+### “Leak”-based approaches
+
+If we’ve decided we don’t want to copy data, another option would be to teach the current persist implementation to designate certain rollups as “backup” rollups, and avoid deleting them and the files they reference. This is a more invasive change to Persist, but would allow us to restore the chosen backups without having to “undelete” any data.
+
+- This complicates the running system, and introduces some new risk: I could easily introduce a bug that caused Persist GC to incorrectly delete blobs that were part of a backup.
+    - Other approaches don’t run this risk: in the copying-based approach, a production environment should never need to delete from the backup location, only write to it; if using S3 versioning, our production environments don’t need permissions to read or modify anything but the most current version.
+
+## Alternatives to whole-environment backups
+
+### Walking the dependency graph in `environmentd`
+
+Imagine a fictional controller that:
+
+- manages all persistent state in an environment, including the stash;
+- tracks the dependencies between all collections, even across intermediate non-durable state like indices;
+- has the ability to hold back the since on shards it controls.
+
+In this world, a toy backup algorithm for a whole Materialize environment might look like:
+
+- Add the stash to the backup.
+- Enumerate all the shards in the environment, according to the catalog as of the moment it was backed up.
+- For every shard S, in reverse dependency order (ie. if A depends on B, process A first):
```

Comment:
extra-nit: In the "Correctness" section, the example is B depends on A. Maybe make this consistent with that?